presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Memo.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.OutputNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+public class Memo
+{
+    private final PlanNodeIdAllocator idAllocator;
+    private final int rootGroup;
+
+    private int nextGroupId;
+
+    private final Map<Integer, PlanNode> membership = new HashMap<>();
+    private final Map<Integer, Integer> referenceCounts = new HashMap<>();
+
+    public Memo(PlanNodeIdAllocator idAllocator, PlanNode plan)
+    {
+        this.idAllocator = idAllocator;
+        rootGroup = insertRecursive(plan);
+        referenceCounts.put(rootGroup, 1);
+    }
+
+    public int getRootGroup()
+    {
+        return rootGroup;
+    }
+
+    public PlanNode getNode(int group)
+    {
+        checkArgument(membership.containsKey(group), "Invalid group: %s", group);
+        return membership.get(group);
+    }
+
+    public PlanNode extract()
+    {
+        return extract(getNode(rootGroup));
+    }
+
+    private PlanNode extract(PlanNode node)
+    {
+        if (node instanceof GroupReference) {
+            return extract(membership.get(((GroupReference) node).getGroupId()));
+        }
+
+        List<PlanNode> children = node.getSources().stream()
+                .map(this::extract)
+                .collect(Collectors.toList());
+
+        return ChildReplacer.replaceChildren(node, children);
+    }
+
+    public void replace(int group, PlanNode node, String rule)
+    {
+        PlanNode old = membership.get(group);
+
+        // output node is special. TODO: we need to get rid of OutputNode. It doesn't add any value
```

Comment:
Sometimes it may a value to keep such nodes as `OutputNode`. For example in case when optimizer optimizes a batch of queries (not just a single query) then it might use such dummy node to group queries together.

Just a note. 

Anyway I would try to handle `OutputNode` as any other node, it is just a special `ProjectNode`.